### PR TITLE
New resource rosa classic

### PR DESCRIPTION
**Pull request by bardielle**
* Added a new resource for rosa STS called ocm_cluster_rosa_classic
* Moving the STS resource to ocm_cluster_rosa_classic resource 
* using an internal module for creating the operator roles (instead of https://github.com/rh-mobb/terraform-aws-rosa-sts-roles )

### CODE REVIEW

**Comment by bardielle** [comment]
@nirarg
@vkareh 
@gdbranco

File: provider/cluster_rosa_classic_state.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/*
```
Comment (nirarg) [inline provider/cluster_rosa_classic_state.go:1]: Suggesting to have this new file under "modules" package and not "provider"
This is an internal module and not provider
Or it can be under "provider/modules"

File: examples/create_rosa_cluster/create_rosa_sts_cluster/main.tf [hunk 2]
```diff
@@ -100,11 +98,11 @@ resource "ocm_cluster" "rosa_cluster" {
...
 
-module sts_roles {
-    source  = "rh-mobb/rosa-sts-roles/aws"
-    create_account_roles = false
-    clusters = [{
-        id = ocm_cluster.rosa_cluster.id
-        operator_role_prefix = var.operator_role_prefix
-    }]
+module operator_roles {
+    source  = "https://github.com/bardielle/terraform-provider-ocm/tree/new_resource_rosa_classic/provider/ocm_cluster_rosa_classic/iam_roles_module"
```
Comment (nirarg) [inline examples/create_rosa_cluster/create_rosa_sts_cluster/main.tf:102]: Make sure to change it to the origin github path (`github.com/openshify-online/...`) before merge

File: modules/ocm_cluster_rosa_classic/README.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+* AWS Admin Account
+* OCM Account and OCM CLI
+* ROSA CLI
+
+## Get OCM Information
+
+When create OCM roles, the roles require org id and external id.
+
+When create User role. the roles require user id and username.
```
Comment (gdbranco) [inline modules/ocm_cluster_rosa_classic/README.md:32]: creating

File: modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf [hunk 4]
```diff
@@ -0,0 +1,32 @@
+resource "aws_iam_role" "cloud-credential_role" {
+  name = "${var.operator_role_prefix}-openshift-cloud-credential-operator-cloud-c"
```
Comment (gdbranco) [inline modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf:2]: "A name for the IAM role, up to 64 characters in length." [1](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html)

What we do the in the CLI in setup the whole name and truncate it to first 64. The ARN does not have this restriction so it will have the whole string

Should that truncate operation be needed here?

File: modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf [hunk 5]
```diff
@@ -0,0 +1,32 @@
+resource "aws_iam_role" "cloud-credential_role" {
```
Comment (gdbranco) [inline modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf:1]: By having the operator roles trust statically setup in these tf files will require us to update this everytime we need a new role, correct?

Could there be some way to generate these dynamically? Also depending on the version chosen for the cluster there are different operator roles not sure if I got how that is being handled

File: modules/ocm_cluster_rosa_classic/operator_roles/cloud_network_config_role.tf [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    red-hat-managed = true
+    rosa_cluster_id = var.cluster_id
+    operator_namespace = "openshift-cloud-network-config-controller"
+    operator_name = "cloud-credentials"
+  }
+}
+
+resource "aws_iam_role_policy_attachment" "cloud_network_config_role_policy_attachment" {
+  role = aws_iam_role.cloud_network_config_role.name
+  policy_arn = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:policy/ManagedOpenShift-openshift-cloud-network-config-controller-cloud"
```
Comment (gdbranco) [inline modules/ocm_cluster_rosa_classic/operator_roles/cloud_network_config_role.tf:31]: The operator policy is not fixed with ManagedOpenshift as the prefix in the CLI, not sure if this will be the case for tf

**Comment by gdbranco** [comment]
There is an empty output.tf file, is this needed?

File: modules/ocm_cluster_rosa_classic/operator_roles/image_registry_role.tf [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    red-hat-managed = true
+    rosa_cluster_id = var.cluster_id
+    operator_namespace = "openshift-image-registry"
+    operator_name = "installer-cloud-credentials"
+  }
+}
+
+resource "aws_iam_role_policy_attachment" "image_registry_role_policy_attachment" {
+  role = aws_iam_role.image_registry_role.name
+  policy_arn = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:policy/ManagedOpenShift-openshift-image-registry-installer-cloud-creden"
```
Comment (oriAdler) [inline modules/ocm_cluster_rosa_classic/operator_roles/image_registry_role.tf:34]:
```suggestion
  policy_arn = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:policy/${var.operator_role_prefix}-openshift-image-registry-installer-cloud-creden"
```

**Review by oriAdler — COMMENTED** [review]
Do we need this file `modules/ocm_cluster_rosa_classic/operator_roles/output.tf`?

I think `modules/ocm_cluster_rosa_classic/operator_roles/variables.tf` and `modules/ocm_cluster_rosa_classic/variables.tf` are identical, do we need the two of them?

File: modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    red-hat-managed = true
+    rosa_cluster_id = var.cluster_id
+    operator_namespace = "openshift-cluster-csi-drivers"
+    operator_name = "ebs-cloud-credentials"
+  }
+}
+
+resource "aws_iam_role_policy_attachment" "csi_drivers_role_policy_attachment" {
+  role = aws_iam_role.csi_drivers_role.name
+  policy_arn = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:policy/ManagedOpenShift-openshift-cluster-csi-drivers-ebs-cloud-credent"
```
[the remark this answers is not in the document]
Comment (gdbranco) [reply modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf:34]: So this might be that eventually, we are actually going over this atm in the CLI as well. Today we use the account roles prefix instead of the operator roles prefix for the operator policy, so hold on to this specific change for a moment

[the remark this answers is not in the document]
Comment (oriAdler) [reply modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf:34]: @gdbranco good point, so I assume it is supposed to be the account role prefix as of now.

@bardielle the `ManagedOpenShift` is the default option for the account roles prefix, 
but the user can choose another value for this prefix.

File: modules/ocm_cluster_rosa_classic/README.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    "updated_at": "2022-06-16T14:17:02Z"
+  },
+  "updated_at": "2022-05-25T02:02:02Z",
+  "username": "shading_mobb"
+}
+```
+
+## Get Clusters Information.
+
+In order to create operator roles for clusters. Users need to provide cluster id, operator role prefix and OIDC Endpoint URL
```
[the remark this answers is not in the document]
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/README.md:62]: I updated it thanks.
Please - rereview

File: modules/ocm_cluster_rosa_classic/README.md (hunk 3 above)
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/README.md:32]: I updated it thanks.
Please - rereview

File: modules/ocm_cluster_rosa_classic/README.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+## Prerequisites
+
+* AWS Admin Account
+* OCM Account and OCM CLI
+* ROSA CLI
+
+## Get OCM Information
+
+When create OCM roles, the roles require org id and external id.
```
[the remark this answers is not in the document]
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/README.md:30]: I updated it thanks.
Please - rereview

File: modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf (hunk 5 above)
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf:1]: @gdbranco for the new role we will need to add one more `tf` file that includes the new IAM role resource.
According to your last question - only the policies can be changed between cluster version (as I understood). when I asked about it in the sync meeting the answer was that roles cannot be changed

File: modules/ocm_cluster_rosa_classic/operator_roles/cloud_network_config_role.tf (hunk 6 above)
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/operator_roles/cloud_network_config_role.tf:31]: Fixed it to ba assigned by a variable

File: modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf (hunk 4 above)
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/operator_roles/cloud_credential_role.tf:2]: My plan is to use only the operator_role_prefix and send the name explicitly. 
This will be part of the next PR

File: modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf:34]: Changed it to be assigned by a new variable called var.account_role_prefix

File: provider/cluster_rosa_classic_state.go (hunk 1 above)
Comment (bardielle) [reply provider/cluster_rosa_classic_state.go:1]: I changed it to be in a different directory

File: modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/operator_roles/csi_driver_role.tf]: I found [this function](https://github.com/openshift/rosa/blob/510a8c7e2c25b85a2ec5e30be3c0deeaa5d44ac9/pkg/aws/helpers.go#L369) for calculating the policy name

**Comment by gdbranco** [comment]
Considering this is a first milestone and more work will still be put into I think it looks good to me

/lgtm

File: provider/cluster_rosa_classic_resource.go [hunk 11]
```diff
@@ -0,0 +1,823 @@
...
+			Value: oidc_endpoint_url,
+		}
+		state.Sts.RoleARN = types.String{
+			Value: sts.RoleARN(),
+		}
+		state.Sts.SupportRoleArn = types.String{
+			Value: sts.SupportRoleARN(),
+		}
+		state.Sts.InstanceIAMRoles.MasterRoleARN = types.String{
+			Value: sts.InstanceIAMRoles().MasterRoleARN(),
```
Comment (nirarg) [inline provider/cluster_rosa_classic_resource.go:680]: Can InstanceIAMRoles return nil?

File: provider/cluster_rosa_classic_resource.go [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+		azs := make([]string, 0)
+		for _, e := range state.AvailabilityZones.Elems {
+			azs = append(azs, e.(types.String).Value)
+		}
+		nodes.AvailabilityZones(azs...)
+	}
+
+	if !state.AutoScalingEnabled.Unknown && !state.AutoScalingEnabled.Null && state.AutoScalingEnabled.Value {
+		autoscaling := cmv1.NewMachinePoolAutoscaling()
+		if !state.MaxReplicas.Unknown && !state.MaxReplicas.Null {
```
Comment (nirarg) [inline provider/cluster_rosa_classic_resource.go:306]: Consider using `ShouldPatchInt` helper functions
Relevant also for all terraform values check here

File: provider/sts.go [hunk 13]
```diff
@@ -5,13 +5,17 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
-func stsResource() tfsdk.NestedAttributes {
+func StsResource() tfsdk.NestedAttributes {
```
Comment (nirarg) [inline provider/sts.go:8]: Why did you change it, its used only from this package

File: provider/helpers.go [hunk 14]
```diff
@@ -20,9 +20,9 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
-// shouldPatchInt changed checks if the change between the given state and plan requires sending a
+// ShouldPatchInt changed checks if the change between the given state and plan requires sending a
 // patch request to the server. If it does it returns the value to add to the patch.
-func shouldPatchInt(state, plan types.Int64) (value int64, ok bool) {
+func ShouldPatchInt(state, plan types.Int64) (value int64, ok bool) {
```
Comment (nirarg) [inline provider/helpers.go:25]: Why did you change it, its used only from this package

File: examples/create_rosa_cluster/create_rosa_sts_cluster/main.tf [hunk 15]
```diff
@@ -36,6 +36,10 @@ variable operator_role_prefix {
     type = string
 }
 
+variable account_role_prefix {
```
[the remark this answers is not in the document]
Comment (bardielle) [reply examples/create_rosa_cluster/create_rosa_sts_cluster/main.tf:39]: I can change the variable name - account_role_prefix_for_operator_role_creation . what do you say?

File: modules/ocm_cluster_rosa_classic/main.tf [hunk 16]
```diff
@@ -0,0 +1,16 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 4.0"
+    }
+  }
+}
+
+module operator_role {
```
[the remark this answers is not in the document]
Comment (bardielle) [reply modules/ocm_cluster_rosa_classic/main.tf]: I change the tree to be : 
`modules/operator_roles`

File: provider/helpers.go (hunk 14 above)
Comment (bardielle) [reply provider/helpers.go:25]: I tried to create a new package but removed it and didn't change it back to be in the original name

File: provider/cluster_rosa_classic_resource.go (hunk 12 above)
Comment (bardielle) [reply provider/cluster_rosa_classic_resource.go:306]: Not sure it will be clear to use a function for that ...
also in other places in the code, it's already uses that check

File: provider/cluster_rosa_classic_resource.go (hunk 11 above)
Comment (bardielle) [reply provider/cluster_rosa_classic_resource.go:680]: fixed

File: provider/cluster_rosa_classic_resource.go [hunk 17]
```diff
@@ -0,0 +1,823 @@
...
+			"external_id": {
+				Description: "Unique external identifier of the cluster.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+			"name": {
+				Description: "Name of the cluster.",
+				Type:        types.StringType,
+				Required:    true,
+			},
```
[the remark this answers is not in the document]
Comment (bardielle) [reply provider/cluster_rosa_classic_resource.go:67]: Added a ticket

File: examples/create_rosa_cluster/create_rosa_sts_cluster/main.tf [hunk 18]
```diff
@@ -100,11 +102,12 @@ resource "ocm_cluster" "rosa_cluster" {
...
 
-module sts_roles {
-    source  = "rh-mobb/rosa-sts-roles/aws"
-    create_account_roles = false
-    clusters = [{
-        id = ocm_cluster.rosa_cluster.id
-        operator_role_prefix = var.operator_role_prefix
-    }]
+module operator_roles {
+    source  = "https://github.com/openshift-online/terraform-provider-ocm/modules/ocm_cluster_rosa_classic/operator_roles"
```
[the remark this answers is not in the document]
Comment (bardielle) [reply examples/create_rosa_cluster/create_rosa_sts_cluster/main.tf:94]: Changed it
`source  = "git::https://github.com/openshift-online/terraform-provider-ocm.git//modules/operator_roles"`

File: provider/cluster_rosa_classic_resource.go [hunk 19]
```diff
@@ -0,0 +1,823 @@
...
+	"net/url"
+	"strings"
+)
+
+const (
+	awsCloudProvider = "aws"
+	rosaProduct      = "rosa"
+)
+
+type ClusterRosaClassicResourceType struct {
```
[the remark this answers is not in the document]
Comment (bardielle) [reply provider/cluster_rosa_classic_resource.go:40]: I tired but it wasn't work - the resource files should be under the same package as the provider, otherwise there is a cycling import

File: provider/cluster_rosa_classic_resource.go [hunk 20]
```diff
@@ -0,0 +1,823 @@
...
+
+type ClusterRosaClassicResource struct {
+	logger     logging.Logger
+	collection *cmv1.ClustersClient
+}
+
+func (t *ClusterRosaClassicResourceType) GetSchema(ctx context.Context) (result tfsdk.Schema,
+	diags diag.Diagnostics) {
+	result = tfsdk.Schema{
+		Description: "OpenShift managed cluster.",
```
[the remark this answers is not in the document]
Comment (bardielle) [reply provider/cluster_rosa_classic_resource.go:51]: Changed it to be `Description: "OpenShift managed cluster using rosa sts.",`
